processing/src/main/java/org/apache/druid/segment/RestrictedSegment.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.segment;
+
+import org.apache.druid.query.filter.DimFilter;
+
+import javax.annotation.Nullable;
+
+public class RestrictedSegment extends WrappedSegmentReference
```

Comment:
Please include javadoc describing the purpose of this class, and how its defensive mechanisms work.

The way they should work is something like: that if you call `asCursorFactory` or `as(CursorFactory.class)` (plus perhaps some other small list), restrictions are handled automatically. But if you call `asQueryableIndex` or `as` for something other than that small list, the query gets the unrestricted internal object and it needs to call some method on the `RestrictedSegment` confirming that it applied the restrictions on its own.

Reply:
> needs to call some method on the RestrictedSegment confirming that it applied the restrictions on its own

To make this robust, we should actually have a 1-1 relationship between `as` calls and "i know what i'm doing" calls. If a query engine calls `as` three different times it should call "i know what i'm doing" three times.